demos/programs/todo/todo.man
.\" $XConsortium: todo.man /main/4 1995/07/17 10:49:07 drk $
.\" Motif
.\"
.\" Copyright (c) 1987-2012, The Open Group. All rights reserved.
.\"
.\" These libraries and programs are free software; you can
.\" redistribute them and/or modify them under the terms of the GNU
.\" Lesser General Public License as published by the Free Software
.\" Foundation; either version 2.1 of the License, or (at your option)
.\" any later version.
.\"
.\" These libraries and programs are distributed in the hope that
.\" they will be useful, but WITHOUT ANY WARRANTY; without even the
.\" implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
.\" PURPOSE. See the GNU Lesser General Public License for more
.\" details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with these librararies and programs; if not, write
.\" to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
.\" Floor, Boston, MA 02110-1301 USA
...\" 
...\" 
...\" HISTORY
.TH todo 1X MOTIF "Demonstration programs"
.SH NAME
\*Ltodo\*O \(em Manage todo lists
.SH SYNOPSIS
.sS
\*Ltodo [-todoFile pathname] \*O
.sE
.SH DESCRIPTION
\*LTodo\*O
is a simple todo list manager.  It provides a demonstration of the
XmNotebook widget.  To use 
\*Ltodo\*O,  you should be able to simply run it
from any directory.  By default it reads its database from your
home directory,  but this can be overriden by \fIpathname\fR.
.SH Files
.nf
.ta 1.2i
$HOME/.todo	The user's todo list
.fi
.SH X Defaults
The class of this application is \fIXmDTodo\fR.  The name of the widgets
of particular interest are:

.nf
.ta 1.2i
notebook	The XmNotebook widget
text	The XmText widget
label	The status area XmLabel widget
.fi

To set a specific resource in your \fI.Xdefaults\fR file,  you specify the
particular widget and resource.  For example,  if we wished to assign specific
colors to the text widget's and notebook widget's background, we could
do the following:

.nf
.ta 3i
XmDTodo*notebook.background:	forestgreen
XmDTodo*text.background:	white
.fi

Additionally there is an application resource to set the default
todo file location:

.nf
.ta 3i
XmDTodo*todoFile:	/users/myhomedir/sometodofile
.fi

.SH BUGS

Todo can occationally get confused when pages are deleted.  Save the current
contents and reopen to resync the application in this case. New pages are
always prepended before the current page,  there should be a way to choose.
Ideally the display should use an XmCSText with rich text and not plain
ascii,  but that went beyond the scope of the demonstration.


.attic/ChangeLog
2017-11-10 11:49:10 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1665 (HEAD, origin/master, origin/HEAD, master)

2017-10-30 23:04:15 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1651

2017-10-30 22:22:31 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1650

2017-10-30 21:47:21 +0200 Oleksiy Chernyavskyy 

	* Fixed bug 1652

2017-08-31 06:28:27 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1655

2017-08-28 05:33:53 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1625

2017-08-26 06:12:14 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1656

2017-08-23 04:10:55 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1658

2017-08-22 06:05:00 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1659

2017-08-22 03:27:36 +0300 Oleksiy Chernyavskyy 

	* Fixed bug 1660

2017-03-28 01:40:33 +0300 Oleksiy Chernyavskyy 

	* updated version and release data

2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3


demos/programs/todo/todo.c
/**
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 */

#ifdef REV_INFO
#ifndef lint
static char *rcsid = "$XConsortium: todo.c /main/6 1995/07/14 09:46:43 drk $";
#endif
#endif

#include <pwd.h>
#include <unistd.h>
#include <stdlib.h>
/* X include files */
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xatom.h>
/* Motif include files */
#include <Xm/Xm.h>
#include <Xm/CascadeBG.h>
#include <Xm/FileSB.h>
#include <Xm/Label.h>
#include <Xm/MainW.h>
#include <Xm/Notebook.h>
#include <Xm/PushBG.h>
#include <Xm/RowColumn.h>
#include <Xm/ScrolledW.h>
#include <Xm/Separator.h>
#include <Xm/Text.h>
/* Demo include files */
#include <Xmd/Help.h>
#include <Xmd/Menus.h>
#include <Xmd/Print.h>

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#if HAVE_STDINT_H
#include <stdint.h>
#elif HAVE_INTTYPES_H
#include <inttypes.h>
#endif

#include "page.h"

#define APP_CLASS "XmdTodo"

#define MAX(x,y) ((x) > (y) ? (x) : (y))
#define MIN(x,y) ((x) > (y) ? (y) : (x))

static char * fallback_resources[] = {
"*text.rows: 24",
"*text.columns: 80",
"*print_manager.printerList: lp,./todo.txt",
"*help_manager.helpFile: todo",
"*notebook.frameShadowThickness: 2",
"*notebook.bindingType:	XmSPIRAL",
NULL
};

/* Options */
OptionsRec options;

#define Offset(field) XtOffsetOf(OptionsRec, field)

XtResource resources[] = {
  {"todoFile", "TodoFile", XtRString, sizeof(String),
    Offset(todoFile), XtRImmediate, NULL}
};

#undef Offset

XrmOptionDescRec optionDesc[] = {
  {"-todoFile", "*todoFile", XrmoptionSepArg, (XtPointer) NULL}
};

static void QuitAppl(Widget, char *, XmPushButtonCallbackStruct *);
static void TextChanged(Widget, XtPointer, XtPointer);

void manageCB(Widget, Widget, XtPointer);
void PresentFDialog(Widget, XtPointer, XmPushButtonCallbackStruct*);
void New(Widget, char*, XmPushButtonCallbackStruct *);
void Save(Widget, char*, XmFileSelectionBoxCallbackStruct *);
void Open(Widget, char*, XmFileSelectionBoxCallbackStruct *);
void Print(Widget, char*, XmdPrintCallbackStruct *);
void SaveIt(Widget, char*, XmPushButtonCallbackStruct *);
void PageChange(Widget, XtPointer, XmNotebookCallbackStruct *);
void NewPage(Widget, XtPointer, XmPushButtonCallbackStruct *);
void DeletePage(Widget, XtPointer, XmPushButtonCallbackStruct *);
void EditPage(Widget, XtPointer, XmPushButtonCallbackStruct *);
void SetPage(int);
void help_cb(Widget, XtPointer, XmAnyCallbackStruct *);

extern void ReadDB(char*);
extern void SaveDB(char*);

extern Page pages[];

Widget shell, notebook, textw, labelw;
Widget help_widget, file_dialog;
int currentPage = 1;
int modified;
extern int maxpages;
struct passwd *user;

int main(int argc, char* argv[])
{
  Widget mainw, menubar;
  Widget *file_menu, *edit_menu, *selected_menu, *help_menu, temp;
  Widget print_widget;
  Cardinal size;
  XtAppContext context;
  Arg args[10];
  int n, i;
  XmString tmp;
  char temppath[256];

  if (argc == 2 && strcmp(argv[1], "-help") == 0) {
    printf("Usage: todo [-todoFile pathname]\n");
    return(0);
  }

  shell		= XtVaAppInitialize(&context, APP_CLASS,
				    optionDesc, XtNumber(optionDesc),
				    &argc, argv,
				    fallback_resources, NULL);

  XtGetApplicationResources(shell, (XtPointer) &options,
			    resources, XtNumber(resources), (Arg *) NULL, 0);

  mainw		= XmCreateMainWindow(shell, "mainw", NULL, 0);
  XtManageChild(mainw);

  menubar	= XmCreateMenuBar(mainw, "menub", NULL, 0);
  XtManageChild(menubar);

  n = 0;
  XtSetArg(args[n], XmNcurrentPageNumber, 1); n++;
  XtSetArg(args[n], XmNlastPageNumber, 100); n++;
  notebook	= XmCreateNotebook(mainw, "notebook", args, n);
  XtManageChild(notebook);
  XtAddCallback(notebook, XmNpageChangedCallback,
		(XtCallbackProc) PageChange, NULL);

  n = 0;
  XtSetArg(args[n], XmNmenuBar, menubar); n++;
  XtSetValues(mainw, args, n);

  n = 0;
  XtSetArg(args[n], XmNpageNumber, 1); n++;
  XtSetArg(args[n], XmNeditMode, XmMULTI_LINE_EDIT); n++;
  textw		= XmCreateScrolledText(notebook, "text", args, n);
  XtManageChild(textw);
  XtAddCallback(textw, XmNvalueChangedCallback, TextChanged, NULL);

  n = 0;
  XtSetArg(args[n], XmNnotebookChildType, XmSTATUS_AREA); n++;
  XtSetArg(args[n], XmNpageNumber, 1); n++;
  labelw	= XmCreateLabel(notebook, "label", args, n);
  XtManageChild(labelw);

  XmdCreateMenu(FILE_MENU, menubar, &file_menu, &size);
  XtAddCallback(file_menu[FILE_NEW], XmNactivateCallback,
		(XtCallbackProc) New, NULL);
  XtAddCallback(file_menu[FILE_OPEN], XmNactivateCallback,
		(XtCallbackProc) PresentFDialog, (XtPointer) Open);
  XtAddCallback(file_menu[FILE_EXIT], XmNactivateCallback,
		(XtCallbackProc) QuitAppl, NULL);
  XtAddCallback(file_menu[FILE_SAVE_AS], XmNactivateCallback,
		(XtCallbackProc) PresentFDialog, (XtPointer) Save);
  XtAddCallback(file_menu[FILE_SAVE], XmNactivateCallback,
		(XtCallbackProc) SaveIt, NULL);

  XmdCreateMenu(SELECTED_MENU, menubar, &selected_menu, &size);
  XtUnmanageChildren(selected_menu, size);
  XtManageChild(selected_menu[SELECTED_PROPERTIES]);
  XtAddCallback(selected_menu[SELECTED_PROPERTIES],
		XmNactivateCallback, (XtCallbackProc) EditPage, NULL);
  XtManageChild(selected_menu[SELECTED_NEW]);
  XtAddCallback(selected_menu[SELECTED_NEW],
		XmNactivateCallback, (XtCallbackProc) NewPage, NULL);
  XtManageChild(selected_menu[SELECTED_DELETE]);
  XtAddCallback(selected_menu[SELECTED_DELETE],
  		XmNactivateCallback, (XtCallbackProc) DeletePage, NULL);

  XmdCreateMenu(HELP_MENU, menubar, &help_menu, &size);
  XtUnmanageChildren(help_menu, size);
  XtManageChild(help_menu[HELP_OVERVIEW]);

  print_widget = XmdCreatePrintDialog(shell, "print_manager", NULL, 0);
  XtAddCallback(print_widget, XmdNprintCallback, (XtCallbackProc) Print, NULL);
  tmp = XmStringCreateLocalized("About Printing");

  XtAddCallback(file_menu[FILE_PRINT], XmNactivateCallback,
		(XtCallbackProc) manageCB, (XtPointer) print_widget);

  help_widget = XmdCreateHelpDialog(shell, "help_manager", NULL, 0);
  XtAddCallback(help_menu[HELP_OVERVIEW], XmNactivateCallback,
		(XtCallbackProc) help_cb, (XtPointer) 0);
  XtAddCallback(print_widget, XmNhelpCallback,
		(XtCallbackProc) help_cb, (XtPointer) 1);

  n = 0;
  XtSetArg(args[n], XmNautoUnmanage, True); n++;
  file_dialog = XmCreateFileSelectionDialog(shell, "file_dialog", args, n);

  user = getpwuid(getuid());
  for (i = 0; i < MAXPAGES; i++) {
    pages[i] = NULL;
  }

  if (options.todoFile == NULL) {
    strcpy(temppath, user -> pw_dir);
    strcat(temppath, "/.todo");
    options.todoFile = XtNewString(temppath);
  } else {
    /* Copy the string for consistency */
    options.todoFile = XtNewString(options.todoFile);
  }

  XtVaSetValues(shell, XmNtitle, options.todoFile,
		XmNtitleEncoding, XA_STRING, NULL, NULL);
  ReadDB(options.todoFile);
  SetPage(0);

  XtRealizeWidget(shell);
  XtAppMainLoop(context);
  return 0;    /* make compiler happy */
}

static void QuitAppl(Widget w, char *i, XmPushButtonCallbackStruct *e)
{
	exit(EXIT_SUCCESS);
}

static void TextChanged(Widget w, XtPointer i, XtPointer i2)
{
	modified = 1;
}

void manageCB(Widget widget, Widget w_to_manage, XtPointer callback)
{
	(void)callback;
	if (w_to_manage)
		XtManageChild(w_to_manage);
}

void
PresentFDialog(Widget w, XtPointer cb, XmPushButtonCallbackStruct *cs)
{
  XtRemoveAllCallbacks(file_dialog, XmNokCallback);
  XtAddCallback(file_dialog, XmNokCallback, (XtCallbackProc) cb, NULL);
  XtManageChild(file_dialog);
}

void
New(Widget w, char* ignore, XmPushButtonCallbackStruct *cs)
{
  char buf[128];
  char *str;
  Boolean found = False;
  int i = 0;

  while(! found) {
    sprintf(buf, "untitled%d.todo", i++);
    found = access(buf, F_OK) != 0;
  }

  str = XtNewString(buf);
  ReadDB(str);
  XtFree(options.todoFile);
  options.todoFile = str;
  XtVaSetValues(shell, XmNtitle, str, XmNtitleEncoding,
		XA_STRING, NULL, NULL);
  SetPage(0);
}

void
Save(Widget w, char *i, XmFileSelectionBoxCallbackStruct *i2)
{
  char *str;
  if ((str = XmStringUnparse(i2->value, NULL, XmCHARSET_TEXT,
			    XmCHARSET_TEXT, NULL, 0, XmOUTPUT_ALL))) {
    SaveDB(str);
    XtFree(options.todoFile);
    options.todoFile = str;
    XtVaSetValues(shell, XmNtitle, str, XmNtitleEncoding,
		  XA_STRING, NULL, NULL);
  }
}

void
Open(Widget w, char *i, XmFileSelectionBoxCallbackStruct *i2)
{
  char *str;
  if ((str = XmStringUnparse(i2->value, NULL, XmCHARSET_TEXT,
			    XmCHARSET_TEXT, NULL, 0, XmOUTPUT_ALL))) {
    ReadDB(str);
    XtFree(options.todoFile);
    options.todoFile = str;
    XtVaSetValues(shell, XmNtitle, str, XmNtitleEncoding,
		  XA_STRING, NULL, NULL);
  }
}

void
help_cb(Widget w, XtPointer item, XmAnyCallbackStruct *cb)
{
  XtManageChild(help_widget);
  XmdGotoHelpItem(w, (intptr_t)item, help_widget);
}

void
Print(Widget w, char *ignore, XmdPrintCallbackStruct *cb)
{
  int i;
  FILE *temp;
  int from, to;

  temp = fopen("/tmp/.todoout", "w");

  if (cb -> first == cb -> last &&
      cb -> first == 0) {
    from = 0;
    to = maxpages - 1;
  } else {
    from = MAX(0, cb -> first - 1);
    to = MIN(maxpages, cb -> last - 1);
  }

  for (i = from; i <= to; i++) {
    if (pages[i] -> label != NULL) {
      fprintf(temp, "Subject: %s\n", pages[i] -> label);
      fprintf(temp, "---------------------------\n\n\n");
    }
    fprintf(temp, "%s", pages[i] -> page);
    if (i != (maxpages - 1)) fprintf(temp, "\f");
  }

  fclose(temp);
  XmdPrintDocument("/tmp/.todoout", cb);
}
